### PR TITLE
Fix watchOS minimum deployment and deprecation of OSAtomicIncrement32Barrier

### LINKED ISSUE #1752: Update WATCHOS_DEPLOYMENT_TARGET to 3.0 for Submission to Apple
**Short description of the issue**:
With the final release of Xcode 10 and watchOS 5, Apple is rejecting submission of watchOS applications with a `WATCHOS_DEPLOYMENT_TARGET` below **3.0**

**Expected outcome**:

Apple accepts the application via upload through the normal means.

**What actually happens**:

The application upload is rejected with the following message:
```
[02:22:23]: [iTMSTransporter] [2018-09-25 02:22:23 MDT] <main> ERROR: ERROR ITMS-90081: "This bundle is invalid. Applications built for more than one architecture require an iOS Deployment Target of 3.0 or later."
```
While the error message suggests the iOS Deployment target is wrong, it appears this is a misnomer, as our iOS deployment target is 9.0. Others are seeing this issue as well:
https://forums.developer.apple.com/thread/108853

We did have to update the watchOS deployment target from 2.0 (the default) on other frameworks we use. We had been relying on the precompiled version of RxSwift to expedite our build times, but it appears that the current minimum version compiled in is 2.0 as shown here:
https://github.com/ReactiveX/RxSwift/blob/master/Rx.xcodeproj/project.pbxproj#L6428

**Self contained code example that reproduces the issue**:

_The issue isn't specific to any code, but how it's built via the project setting linked above_

**RxSwift/RxCocoa/RxBlocking/RxTest version/commit**
https://github.com/ReactiveX/RxSwift/releases/tag/4.3.1

**Platform/Environment**

  - [ ] iOS
  - [ ] macOS
  - [ ] tvOS
  - [x] watchOS
  - [ ] playgrounds

**How easy is to reproduce? (chances of successful reproduce after running the self contained code)**

  - [x] easy, 100% repro
  - [ ] sometimes, 10%-100%
  - [ ] hard, 2% - 10%
  - [ ] extremely hard, %0 - 2%

**Xcode version**:

```
  10.0 (10A255)
```

:warning: Fields below are optional for general issues or in case those questions aren't related to your issue, but filling them out will increase the chances of getting your issue resolved. :warning:

**Installation method**:
  - [ ] CocoaPods
  - [x] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  (so we can know if this is a potential cause of your issue)
  - [x] yes (which ones: 9.0, 9.1, 9.2, 9.3.1, 10.0)
  - [ ] no

**Level of RxSwift knowledge**:
  (this is so we can understand your level of knowledge
    and formulate the response in an appropriate manner)
  - [ ] just starting
  - [ ] I have a small code base
  - [x] I have a significant code base

### PR DESCRIPTION
**Pull request by freak4pc**
Supersedes #1753.
Fixes #1752. 

This PR:

* Takes @dlewanda's changes for the watchOS 3.0 minimum deployment.
* Fixes the deprecation of `OSAtomicIncrement32Barrier` by switching to `atomic_fetch_add` from `stdatomic.h`. 

As a reminder - we should merge this ASAP as it causes a critical issue when submitting to the App Store with a watchOS target.

### CODE REVIEW

**Comment by freak4pc** [comment]
<p align="center"></p>

**Comment by freak4pc** [comment]
@kzaher Let me know if this makes sense for you

**Comment by freak4pc** [comment]
Hey @kzaher - Just want to clarify a few things:

1. This PR moves all watchOS targets to 3.0, not just the RxCocoa one. 
2. The reason the fix was applied to the RxCocoa repo is because that methods is specifically hard deprecated (with an error, not a warning) in Objective C. 
3. In the Swift-related code this only causes a warning at the moment, not a build error (even for watchOS 3).

This PR should be sufficient to release a patch to resolve the critical issue at hand (can't build for watchOS 3, can't submit to App Store with watchOS 3) for now in my opinion, I think we should wait to see if we're getting a proper public API for stdatomic before we use our own implementation for everything, but that's just my personal opinions. 

tl;dr - 
RxCocoa is a build error (since its Objective-C) - so it was fixed 
RxSwift is only a warning - so I left it as is, for the time being. 

Appreciate your thoughts on this - I'm OK with whatever you decide, clearly.

Shai.

**Comment by kzaher** [comment]
Hi @freak4pc ,

If we are publishing fixes, we should fix these warnings IMHO otherwise we might need another patch release for these warnings.

**Comment by freak4pc** [comment]
I don't super mind but we had many releases with warning for a while and it
was never a issue- since we can't control random depreciation without
fallback from Apple.

I can switch to your implementation but I hope we won't see a performance
hit with that.

-- 
Shai Mishali

**Comment by freak4pc** [comment]
Hey @kzaher - Removing the two build warnings is not enough to make this code work on watchOS because its importing Glibs and XCTest, but even when resolving this it seems that 

`currentThread.threadDictionary = threadDictionary`

doesn't work for the watchOS target (only for the Linux one). Any thoughts?

**Comment by freak4pc** [comment]
Re-ping @kzaher - Without this fix (or forking the repo and applying it), people with watchOS targets can't submit to the App Store (reject).

**Comment by kzaher** [comment]
> Hey @kzaher - Removing the two build warnings is not enough to make this code work on watchOS because its importing Glibs and XCTest, but even when resolving this it seems that

> currentThread.threadDictionary = threadDictionary

> doesn't work for the watchOS target (only for the Linux one). Any thoughts?

Ok, I will fix the RxSwift then.
